src/main/java/util/InitialDataLoadApplication.java

```diff
@@ -35,23 +41,61 @@ public static void main(String[] args) {
-                    // Convert the row into a Monument objects
-                    Monument newMonument = CsvMonumentConverter.convertCsvRowToMonument(csvRow.strip());
-                    // Add the newMonument to the accumulating list
-                    newMonuments.add(newMonument);
+                    // Convert the row into a CsvMonumentConverterResult object
+                    CsvMonumentConverterResult result = CsvMonumentConverter.convertCsvRow(csvRow.strip());
+                    // Validate the CsvMonumentConverterResult's Monument
+                    List<String> violationMessages = result.getMonument().validate();
+                    // If there were no violations, add the CsvMonumentConverterResult to the accumulating list
+                    if (violationMessages == null) {
+                        results.add(result);
+                    }
+                    else {
+                        System.out.println("Failed to validate Monument: " + result.getMonument().toString());
+                        System.out.println("Reasons: ");
+                        for (String violationMessage : violationMessages) {
+                            System.out.println("\t" + violationMessage);
+                        }
+                    }
                 }
                 catch (Exception e) {
                     System.out.println("Failed to process row number: " + rowCount);
                     System.out.println("Exception: " + e.toString());
                 }
             }
 
-            System.out.println("Number of new Monuments created: " + newMonuments.size());
+            System.out.println("Number of Monuments created: " + results.size());
+
+            int monumentInsertedCount = 0;
+            int tagInsertedCount = 0;
+
+            for (CsvMonumentConverterResult r : results) {
 
-            // Insert all accumulated monuments into the database
-            List<Integer> monumentIdsInserted = monumentService.insert(newMonuments);
+                // Insert the Monument
+                monumentService.insert(r.getMonument());
+                monumentInsertedCount++;
+
+                // Insert all of the Tags associated with the Monument
+                // This isn't done by the above insert call because the Tag Model "owns" the monument_tag join table
+                if (r.getTags() != null) {
+                    for (Tag t : r.getTags()) {
+                        try {
+                            tagService.insert(t);
+                            tagInsertedCount++;
+                        } catch (ConstraintViolationException e) {
+                            // If a ConstraintViolationException is caught, there's a duplicate Tag in the list we are trying
+                            // to insert
+                            // Instead, get the ID of the already existing Tag and update it
+                            // Assume there is only 1 Tag with that Name due to the Unique Constraint
```

Comment:
Having to catch a `ConstraintViolationException`, grabbing the duplicate's ID, setting it then updating isn't ideal. Perhaps in the future we can change `insert()` to encapsulate this logic so the caller doesn't have to worry about it.

Reply:
You could override `doInsert` to achieve this pretty easily for one service

Reply:
Don't feel like doing that in this PR, but we should in the future